go/tools/gazelle/packages/fileinfo.go

```diff
@@ -157,6 +138,25 @@ func fileNameInfo(dir, name string) fileInfo {
 		category = ignoredExt
 	}
 
+	// Determine test, goos, and goarch. This is intended to match the logic
+	// in goodOSArchFile in go/build.
+	var isTest bool
+	var goos, goarch string
+	l := strings.Split(name[:len(name)-len(ext)], "_")
+	if len(l) >= 2 && l[len(l)-1] == "test" {
+		isTest = true && category != unsupportedExt && category != ignoredExt
```

Comment:
I think this would be simpler as:

```
if category == goExt && len(l) >= 2 && l[len(l)-1] == "test" {
  isTest = true
}
```

+1 to the last thing Ian said though. Would be good for `Package.addFile` to have an early check that ignores files with `ignoredExt` and `unsupportedExt`.